### PR TITLE
added tasks tests

### PR DESCRIPTION
**Pull request by sychova**

### CODE REVIEW

File: tests/task.test.js [hunk 1]
```diff
@@ -27,6 +27,16 @@ test('Should create task for user', async() => {
     expect(task.completed).toBe(false)
 })
 
+test('Should not create task with invalid title', async() => {
```
Comment (alexandr-senyuk) [inline tests/task.test.js]: add describe section to explain what exactly you test
```javasctip
describe('POST /tasks', () => {
...
```

File: tests/task.test.js [hunk 2]
```diff
@@ -27,6 +27,16 @@ test('Should create task for user', async() => {
...
+    await request(app)
```
Comment (alexandr-senyuk) [inline tests/task.test.js]: code duplication through all test cases 
```javascript
    await request(app)
        .get(`/tasks/${task0._id}`)
        .set('Authorization', `Bearer ${user0.tokens[0].token}`)
        .send()
```

File: tests/task.test.js [hunk 3]
```diff
@@ -36,12 +46,92 @@ test('Should return all tasks of user 0', async() => {
     expect(response.body.length).toEqual(2)
 })
 
-test('Should not allow deleting tasks that the user doesn\'t own', async() => {
+test('Should fetch user task by id', async() => {
```
Comment (alexandr-senyuk) [inline tests/task.test.js]: you should split tests by subjects (different files for different resources and actions (http methods))

Comment (alexandr-senyuk) [inline tests/task.test.js]:
```suggestion
it('fetches user task by id', async () => {
```

File: tests/task.test.js [hunk 4]
```diff
@@ -36,12 +46,92 @@ test('Should return all tasks of user 0', async() => {
...
+})
+
+test('Should not delete task if unauthenticated', async() => {
+    await request(app)
+        .delete(`/tasks/${task1._id}`)
+        .send()
+        .expect(401)
+    const task = await Task.findById(task1._id)
+    expect(task).not.toBeNull()
 })
```
Comment (alexandr-senyuk) [inline tests/task.test.js]: eof

File: tests/task.test.js [hunk 5]
```diff
@@ -36,12 +46,92 @@ test('Should return all tasks of user 0', async() => {
...
+
+test('Should not update task with invalid title', async() => {
+    await request(app)
+        .patch(`/tasks/${task1._id}`)
+        .set('Authorization', `Bearer ${user0.tokens[0].token}`)
+        .send({
+            title: ''
+        }).expect(400)
+    const task = await Task.findById(task1._id)
+    expect(task.title).not.toBeNull()
```
Comment (alexandr-senyuk) [inline tests/task.test.js]: it's broken. empty string is not null as well

File: tests/task.test.js [hunk 6]
```diff
@@ -36,12 +46,92 @@ test('Should return all tasks of user 0', async() => {
...
+})
+
+test('Should fetch only incomplete tasks', async() => {
     const response = await request(app)
+        .get('/tasks?completed=false')
+        .set('Authorization', `Bearer ${user0.tokens[0].token}`)
+        .send()
+        .expect(200)
+    expect(response.body.length).toEqual(1)
+    console.log(response.body)
```
Comment (alexandr-senyuk) [inline tests/task.test.js]: remove

File: tests/task.test.js [hunk 7]
```diff
@@ -36,12 +46,92 @@ test('Should return all tasks of user 0', async() => {
...
+        .expect(404)
+})
+
+test('Should fetch only completed tasks', async() => {
+    const response = await request(app)
+        .get('/tasks?completed=true')
+        .set('Authorization', `Bearer ${user0.tokens[0].token}`)
+        .send()
+        .expect(200)
+    expect(response.body.length).toEqual(1)
```
Comment (alexandr-senyuk) [inline tests/task.test.js]: https://jestjs.io/docs/expect#tohavelengthnumber it provides more detailed failure message

File: tests/task.test.js (hunk 3 above)
Comment (sychova) [reply tests/task.test.js]: Fixed.

File: tests/task.test.js (hunk 4 above)
Comment (sychova) [reply tests/task.test.js]: Fixed.

File: tests/task.test.js (hunk 6 above)
Comment (sychova) [reply tests/task.test.js]: Fixed.

File: tests/task.test.js (hunk 1 above)
Comment (sychova) [reply tests/task.test.js]: Should it specifically be broken down according to request type?
Can it be broken down logically as in Creating/Updating/Deleting?

File: tests/task.test.js (hunk 3 above)
Comment (sychova) [reply tests/task.test.js]: Should it specifically be broken down according to request type?
Can it be broken down logically as in Creating/Updating/Deleting?

File: tests/task.test.js (hunk 2 above)
Comment (sychova) [reply tests/task.test.js]: Could you please clarify what do you mean exactly?
And what do you want me to do here?
I mean the `.get()` part is mostly different between tests. same as `.set()`.

File: tests/task.test.js (hunk 5 above)
Comment (sychova) [reply tests/task.test.js]: Changed to `toBeTruthy()`

File: tests/task.test.js (hunk 7 above)
Comment (sychova) [reply tests/task.test.js]: Fixed.

File: tests/task.test.js (hunk 3 above)
Comment (sychova) [reply tests/task.test.js]: Fixed.

File: tests/task.test.js (hunk 2 above)
Comment (sychova) [reply tests/task.test.js]: Fixed.

File: tests/task.test.js (hunk 1 above)
Comment (sychova) [reply tests/task.test.js]: Fixed.

File: tests/task.test.js (hunk 5 above)
Comment (sychova) [reply tests/task.test.js]: Fixed.

File: tests/task.test.js (hunk 7 above)
Comment (alexandr-senyuk) [reply tests/task.test.js]: not fixed

File: tests/tasks/create.test.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+        it('creates a user task', async () => {
+            const { body } = await subject()
+            task = await Task.findById(body._id)
+            expect(task).not.toBeNull()
+        })
+
+        it('sets default status correctly', async () => {
+            const { body } = await subject()
+            expect(body.completed).toBe(false)
```
Comment (alexandr-senyuk) [inline tests/tasks/create.test.js]: https://jestjs.io/docs/expect#tobefalsy

File: tests/tasks/get.test.js [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        })
+
+        it('returns 200 response status code', async () => {
+            const { status } = await subject()
+            expect(status).toBe(200)
+        })
+
+        it('returns all tasks of the user', async () => {
+            const { body } = await subject()
+            expect(body.length).toEqual(2)
```
Comment (alexandr-senyuk) [inline tests/tasks/get.test.js]: https://jestjs.io/docs/expect#tohavelengthnumber

File: tests/tasks/create.test.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+        it('returns 400 response status code', async () => {
+            const { status } = await subject()
+            expect(status).toBe(400)
+        })
+
+        it('does not create task without title', async () => {
+            await subject()
+            const task = await Task.find({ title: '' })
+            expect(task.length).toEqual(0)
```
Comment (alexandr-senyuk) [inline tests/tasks/create.test.js]: https://jestjs.io/docs/expect#tohavelengthnumber

File: tests/tasks/create.test.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+        it('returns 401 response status code', async () => {
+            const { status } = await subject()
+            expect(status).toBe(401)
+        })
+
+        it('does not create the task', async () => {
+            await subject()
+            task = await Task.find({ createdBy: user.id })
+            expect(task.length).toEqual(0)
```
Comment (alexandr-senyuk) [inline tests/tasks/create.test.js]: https://jestjs.io/docs/expect#tohavelengthnumber

**Review by alexandr-senyuk — CHANGES_REQUESTED** [review]
Squash commits before next review

File: tests/task.test.js (hunk 7 above)
Comment (sychova) [reply tests/task.test.js]: Fixed. (I hope)

File: tests/tasks/create.test.js (hunk 8 above)
Comment (sychova) [reply tests/tasks/create.test.js]: Fixed.

File: tests/tasks/get.test.js (hunk 9 above)
Comment (sychova) [reply tests/tasks/get.test.js]: Fixed.

File: tests/tasks/create.test.js (hunk 10 above)
Comment (sychova) [reply tests/tasks/create.test.js]: Fixed.

File: tests/tasks/create.test.js (hunk 11 above)
Comment (sychova) [reply tests/tasks/create.test.js]: Fixed.

File: tests/fixtures/tasksDB.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
+const mongoose = require('mongoose')
```
Comment (alexandr-senyuk) [inline tests/fixtures/tasksDB.js]: remove unused files
same for usersDB

File: src/middleware/fileUpload.js [hunk 13]
```diff
@@ -1,22 +1,22 @@
...
-      callback(null, true);
-    },
-  });
-};
+    return multer({
+        limits: {
+            fileSize: 1000000,
+        },
+        fileFilter(req, file, callback) {
+            const allowedExtensions = ['.jpg', '.jpeg', '.png']
```
Comment (alexandr-senyuk) [inline src/middleware/fileUpload.js]: define this as a constant (upper snakecase) in the file root

File: tests/fixtures/tasksDB.js (hunk 12 above)
Comment (sychova) [reply tests/fixtures/tasksDB.js]: Fixed.

File: src/middleware/fileUpload.js (hunk 13 above)
Comment (sychova) [reply src/middleware/fileUpload.js]: Fixed.

**Pull request merged**